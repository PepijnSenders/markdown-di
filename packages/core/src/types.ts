import type { z } from 'zod';

/**
 * Validation error types
 */
export interface ValidationError {
  type: 'frontmatter' | 'partial' | 'file' | 'circular' | 'syntax' | 'schema' | 'injection';
  message: string;
  location: string;
  details?: any;
}

/**
 * Processing options
 */
export interface ProcessOptions {
  content: string;
  baseDir: string;
  mode?: 'validate' | 'build';
  currentFile?: string;
}

/**
 * Processing result
 */
export interface ProcessResult {
  content: string;
  frontmatter: FrontmatterData;
  errors: ValidationError[];
  dependencies: string[];
}

/**
 * Frontmatter data structure
 */
export interface FrontmatterData {
  name: string;
  description: string;
  partials?: Record<string, string | string[]>;
  'output-frontmatter'?: string[];
  [key: string]: any;
}

/**
 * Dependency reference
 */
export interface DependencyReference {
  type: 'partial';
  key: string;
  fullPath: string;
  sourceLine: number;
  sourceColumn: number;
}

/**
 * Processing context
 */
export interface ProcessingContext {
  baseDir: string;
  mode: 'validate' | 'build';
  visitedFiles: Set<string>;
  currentFile?: string;
}