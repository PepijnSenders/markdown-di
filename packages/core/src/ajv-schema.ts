import Ajv from 'ajv'
import addFormats from 'ajv-formats'
import type { ValidationError } from './types'

/**
 * AJV-based schema validator using JSON Schema
 * This is the recommended validator for CLI and simple use cases
 */
export class AjvSchemaValidator {
  private ajv: Ajv
  private schemas: Map<string, any> = new Map()

  constructor() {
    this.ajv = new Ajv({
      allErrors: true, // Collect all errors, not just the first
      strict: false, // Allow unknown keywords
      validateFormats: true, // Validate format keywords
    })
    addFormats(this.ajv) // Adds date, email, uri, uuid, etc formats
  }

  /**
   * Register a named schema
   */
  registerSchema(name: string, jsonSchema: any): void {
    this.schemas.set(name, jsonSchema)
  }

  /**
   * Load multiple schemas from config
   */
  loadSchemas(schemas: Record<string, any>): void {
    Object.entries(schemas).forEach(([name, schema]) => {
      this.registerSchema(name, schema)
    })
  }

  /**
   * Get list of registered schema names
   */
  getSchemaNames(): string[] {
    return Array.from(this.schemas.keys())
  }

  /**
   * Validate data against a schema
   * @param data - Data to validate
   * @param schema - Schema name (string) or inline JSON Schema object
   */
  validate(
    data: any,
    schema: string | any,
  ): { valid: boolean; errors: ValidationError[] } {
    let jsonSchema: any

    if (typeof schema === 'string') {
      // Named schema reference
      jsonSchema = this.schemas.get(schema)
      if (!jsonSchema) {
        const available = this.getSchemaNames()
        const suggestion =
          available.length > 0
            ? ` Available schemas: ${available.join(', ')}`
            : ' No schemas registered.'
        return {
          valid: false,
          errors: [
            {
              type: 'schema',
              message: `Schema "${schema}" not found.${suggestion}`,
              location: 'frontmatter.schema',
            },
          ],
        }
      }
    } else {
      // Inline JSON Schema object
      jsonSchema = schema
    }

    const valid = this.ajv.validate(jsonSchema, data)

    if (valid) {
      return { valid: true, errors: [] }
    }

    // Convert AJV errors to our ValidationError format
    const errors: ValidationError[] = (this.ajv.errors || []).map((err) => {
      const path = err.instancePath || 'root'
      const message = err.message || 'Validation failed'
      return {
        type: 'schema',
        message: `${path}: ${message}`,
        location: err.instancePath || 'frontmatter',
      }
    })

    return { valid: false, errors }
  }
}
