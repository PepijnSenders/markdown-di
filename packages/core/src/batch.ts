import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join, relative, resolve } from "node:path";
import { globSync } from "fast-glob";
import type { ProcessOptions, ValidationError, VariantGenerator } from "./types";
import { MarkdownDI } from "./index";

/**
 * Batch processing configuration
 * Extends ProcessOptions with batch-specific options
 */
export interface BatchConfig extends Pick<ProcessOptions, 'baseDir' | 'onBeforeCompile' | 'variants' | 'mustache' | 'validateFrontmatter'> {
  /**
   * Glob patterns to match markdown files
   * @default ['**\/*.md']
   */
  include?: string[];

  /**
   * Glob patterns to exclude
   * @default ['node_modules/**', '.git/**']
   */
  exclude?: string[];

  /**
   * Output directory for processed files
   * If not provided, files are updated in place
   */
  outDir?: string;

  /**
   * Check mode - returns results without writing files
   * @default false
   */
  check?: boolean;

  /**
   * Silent mode - suppress console output
   * @default false
   */
  silent?: boolean;
}

/**
 * Result for a single file
 */
export interface FileResult {
  file: string;
  changed: boolean;
  errors: ValidationError[];
  /**
   * Simple error messages for easy access
   */
  messages: string[];
}

/**
 * Batch processing result
 */
export interface BatchResult {
  /**
   * Total number of files processed
   */
  totalFiles: number;

  /**
   * Number of files that changed
   */
  changedFiles: number;

  /**
   * Total number of errors across all files
   */
  totalErrors: number;

  /**
   * Results for each processed file
   */
  files: FileResult[];

  /**
   * Whether the batch processing succeeded (no errors)
   */
  success: boolean;

  /**
   * All errors grouped by file for easy access
   * Only includes files with errors
   */
  errorsByFile: Record<string, ValidationError[]>;

  /**
   * All errors flattened into a single array for easy access
   */
  allErrors: ValidationError[];

  /**
   * All error messages as simple strings for easy access
   */
  errorMessages: string[];
}

/**
 * Batch processor for multiple markdown files
 */
export class BatchProcessor {
  private mdi: MarkdownDI;
  private config: Required<
    Omit<BatchConfig, "onBeforeCompile" | "outDir" | "variants" | "mustache" | "validateFrontmatter">
  > & {
    outDir?: string;
    onBeforeCompile?: BatchConfig["onBeforeCompile"];
    variants?: BatchConfig["variants"];
    mustache?: BatchConfig["mustache"];
    validateFrontmatter?: BatchConfig["validateFrontmatter"];
  };

  constructor(config: BatchConfig) {
    this.mdi = new MarkdownDI();
    this.config = {
      baseDir: resolve(config.baseDir),
      include: config.include || ["**/*.md"],
      exclude: config.exclude || ["node_modules/**", ".git/**"],
      outDir: config.outDir ? resolve(config.outDir) : undefined,
      check: config.check || false,
      silent: config.silent || false,
      onBeforeCompile: config.onBeforeCompile,
      variants: config.variants,
      mustache: config.mustache,
      validateFrontmatter: config.validateFrontmatter,
    };
  }

  /**
   * Process all matching markdown files
   */
  async process(): Promise<BatchResult> {
    const { baseDir, include, exclude, outDir, check, silent, variants } = this.config;

    // Find all markdown files
    const files = globSync(include, {
      cwd: baseDir,
      absolute: true,
      ignore: exclude,
    });

    if (files.length === 0) {
      if (!silent) {
        console.warn("No markdown files found");
      }
      return {
        totalFiles: 0,
        changedFiles: 0,
        totalErrors: 0,
        files: [],
        success: true,
        errorsByFile: {},
        allErrors: [],
        errorMessages: [],
      };
    }

    const results: FileResult[] = [];
    let changedCount = 0;
    let errorCount = 0;

    for (const file of files) {
      const content = readFileSync(file, "utf-8");
      const relativePath = relative(baseDir, file);

      // First pass: extract frontmatter to detect if this file has variants configured
      // Skip $dynamic validation on this pass - variants will provide the data on second pass
      const result = await this.mdi._processForBatch({
        content,
        baseDir,
        currentFile: file,
        mode: "build",
        onBeforeCompile: this.config.onBeforeCompile,
        mustache: this.config.mustache,
        validateFrontmatter: this.config.validateFrontmatter,
      });

      // Check if this file has variants configured
      const fileId = result.frontmatter.id as string | undefined;
      const variantConfig = fileId && variants ? variants[fileId] : undefined;

      if (variantConfig) {
        // This file has variants - process each variant instead of the original
        for (let i = 0; i < variantConfig.data.length; i++) {
          const variantData = variantConfig.data[i];

          // Re-process the file with variant data
          const variantResult = await this.mdi.process({
            content,
            baseDir,
            currentFile: file,
            mode: "build",
            onBeforeCompile: async (context) => {
              // Merge variant data with any onBeforeCompile results
              const baseData = this.config.onBeforeCompile
                ? await this.config.onBeforeCompile(context)
                : {};
              return { ...baseData, ...variantData };
            },
            mustache: this.config.mustache,
            validateFrontmatter: this.config.validateFrontmatter,
          });

          const variantChanged = variantResult.content !== content;
          if (variantChanged) changedCount++;
          errorCount += variantResult.errors.length;

          // Get output path from callback
          const variantOutputPath = variantConfig.getOutputPath(
            {
              id: fileId!, // fileId is guaranteed to exist here since variantConfig is only set when fileId exists
              filePath: file,
              frontmatter: variantResult.frontmatter as Record<string, unknown>,
              baseDir,
              dynamicFields: [],
            },
            variantData,
            i
          );

          results.push({
            file: variantOutputPath,
            changed: variantChanged,
            errors: variantResult.errors,
            messages: variantResult.errors.map(e => e.message),
          });

          // Write variant file if not in check mode and no errors
          if (!check && variantChanged && variantResult.errors.length === 0) {
            const fullOutputPath = outDir
              ? join(outDir, variantOutputPath)
              : join(baseDir, variantOutputPath);

            const outputDir = dirname(fullOutputPath);
            if (!existsSync(outputDir)) {
              mkdirSync(outputDir, { recursive: true });
            }

            writeFileSync(fullOutputPath, variantResult.content, "utf-8");
          }
        }
      } else {
        // Normal file processing (no variants)
        const changed = result.content !== content;
        if (changed) changedCount++;
        errorCount += result.errors.length;

        results.push({
          file: relativePath,
          changed,
          errors: result.errors,
          messages: result.errors.map(e => e.message),
        });

        // Write files only if:
        // 1. Not in check mode
        // 2. Content changed
        // 3. No validation errors
        if (!check && changed && result.errors.length === 0) {
          const outputPath = outDir ? join(outDir, relativePath) : file;

          // Create output directory if needed
          const outputDir = dirname(outputPath);
          if (!existsSync(outputDir)) {
            mkdirSync(outputDir, { recursive: true });
          }

          writeFileSync(outputPath, result.content, "utf-8");
        }
      }
    }

    // Build error aggregations for easy access
    const errorsByFile: Record<string, ValidationError[]> = {};
    const allErrors: ValidationError[] = [];
    const errorMessages: string[] = [];

    for (const result of results) {
      if (result.errors.length > 0) {
        errorsByFile[result.file] = result.errors;
        allErrors.push(...result.errors);
        errorMessages.push(...result.messages);
      }
    }

    const batchResult: BatchResult = {
      totalFiles: files.length,
      changedFiles: changedCount,
      totalErrors: errorCount,
      files: results,
      success: errorCount === 0,
      errorsByFile,
      allErrors,
      errorMessages,
    };

    // Log results if not silent
    if (!silent) {
      this.logResults(batchResult);
    }

    return batchResult;
  }

  /**
   * Log processing results to console
   */
  private logResults(result: BatchResult): void {
    const { totalFiles, changedFiles, totalErrors, files } = result;

    if (totalErrors > 0) {
      console.error(`✗ Found ${totalErrors} errors in ${totalFiles} files`);
    } else if (this.config.check && changedFiles > 0) {
      console.warn(`${changedFiles} file(s) would be formatted`);
      files.forEach((fileResult) => {
        if (fileResult.changed) {
          console.log(`  ${fileResult.file}`);
        }
      });
    } else if (changedFiles > 0) {
      console.log(`✓ Formatted ${changedFiles} file(s)`);
      if (this.config.outDir) {
        console.log(`  Output: ${this.config.outDir}`);
      }
    } else {
      console.log(`✓ All ${totalFiles} files are correctly formatted`);
    }
  }

}
