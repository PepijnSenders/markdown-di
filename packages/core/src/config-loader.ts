import { existsSync, readFileSync } from 'node:fs'
import { dirname, join } from 'node:path'

export interface MarkdownDIConfig {
  schemas: Record<string, any> // JSON Schema objects
}

/**
 * Loads configuration files for markdown-di
 * Supports auto-discovery by walking up directories
 */
export class ConfigLoader {
  private static readonly CONFIG_NAMES = [
    '.markdown-di.json',
    '.markdown-di.schemas.json',
    'markdown-di.config.json',
  ]

  /**
   * Find config file by walking up directories (like .git, .prettierrc)
   * @param startDir - Directory to start searching from
   * @returns Path to config file or null if not found
   */
  findConfig(startDir: string): string | null {
    let currentDir = startDir
    const root = dirname(currentDir) === currentDir // Reached filesystem root

    while (!root) {
      for (const name of ConfigLoader.CONFIG_NAMES) {
        const configPath = join(currentDir, name)
        if (existsSync(configPath)) {
          return configPath
        }
      }

      const parentDir = dirname(currentDir)
      if (parentDir === currentDir) break // Reached root
      currentDir = parentDir
    }

    return null
  }

  /**
   * Load and parse config file
   * @param configPath - Path to config file
   * @returns Parsed config object
   */
  loadConfig(configPath: string): MarkdownDIConfig {
    if (!existsSync(configPath)) {
      throw new Error(`Config file not found: ${configPath}`)
    }

    const content = readFileSync(configPath, 'utf-8')

    try {
      const parsed = JSON.parse(content)

      // Validate basic structure
      if (!parsed.schemas || typeof parsed.schemas !== 'object') {
        throw new Error('Config must have a "schemas" object')
      }

      return parsed as MarkdownDIConfig
    } catch (err) {
      if (err instanceof SyntaxError) {
        throw new Error(`Invalid JSON in config file: ${err.message}`)
      }
      throw err
    }
  }

  /**
   * Load config with auto-discovery
   * @param startDir - Directory to start searching from
   * @param explicitPath - Optional explicit config path (overrides auto-discovery)
   * @returns Config object or null if not found
   */
  loadConfigAuto(startDir: string, explicitPath?: string): MarkdownDIConfig | null {
    const configPath = explicitPath || this.findConfig(startDir)

    if (!configPath) {
      return null
    }

    return this.loadConfig(configPath)
  }
}
