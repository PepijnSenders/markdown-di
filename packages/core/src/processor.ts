import { readFileSync } from 'node:fs'
import { isDynamicPattern } from 'fast-glob'
import matter from 'gray-matter'
import Mustache from 'mustache'
import type { CircularDependencyDetector, DependencyResolver } from './resolver'
import type { FrontmatterData, ProcessingContext, ValidationError } from './types'

/**
 * Processes markdown content with dependency injection
 */
export class ContentProcessor {
  constructor(
    private resolver: DependencyResolver,
    private circularDetector: CircularDependencyDetector,
  ) {}

  async process(
    content: string,
    frontmatter: FrontmatterData,
    context: ProcessingContext,
  ): Promise<{
    processedContent: string
    errors: ValidationError[]
    dependencies: string[]
  }> {
    const errors: ValidationError[] = []

    // Resolve dependencies and check for issues
    const { dependencies, errors: resolutionErrors } = this.resolver.resolve(frontmatter)
    errors.push(...resolutionErrors)

    // Check for circular dependencies
    if (context.currentFile) {
      const circularErrors = this.circularDetector.detect(context.currentFile, dependencies)
      errors.push(...circularErrors)
    }

    // If in validate mode, return early
    if (context.mode === 'validate') {
      return {
        processedContent: content,
        errors,
        dependencies,
      }
    }

    // Create view object from frontmatter
    const view: any = { ...frontmatter }

    // Resolve partials to their file contents
    // Skip if we already have resolution errors for this partial
    if (frontmatter.partials) {
      view.partials = {}

      for (const [key, value] of Object.entries(frontmatter.partials)) {
        // Check if this partial already has an error from resolution phase
        const hasResolutionError = resolutionErrors.some(
          (err) => err.location === `partials.${key}`,
        )

        if (hasResolutionError) {
          // Set empty string so Mustache doesn't break
          view.partials[key] = ''
          continue
        }

        // Only try to resolve content if there was no resolution error
        const fileContent = await this.resolvePartialContent(key, value, context)
        view.partials[key] = fileContent
      }
    }

    // Use Mustache to render everything in one pass
    let processedContent: string
    try {
      processedContent = Mustache.render(content, view)
    } catch (error) {
      errors.push({
        type: 'injection',
        message: `Mustache templating error: ${error}`,
        location: 'content',
      })
      processedContent = content
    }

    return { processedContent, errors, dependencies }
  }

  /**
   * Resolve a partial definition to its file content
   */
  private async resolvePartialContent(
    _key: string,
    value: string | string[],
    context: ProcessingContext,
  ): Promise<string> {
    const contents: string[] = []

    const patterns = Array.isArray(value) ? value : [value]

    for (const pattern of patterns) {
      // Check if it's a glob pattern using fast-glob's helper
      if (isDynamicPattern(pattern)) {
        // Treat as glob pattern
        const filePaths = this.resolver.resolveGlobPattern(context.baseDir, pattern)
        for (const filePath of filePaths) {
          const content = readFileSync(filePath, 'utf-8')
          contents.push(content)
        }
      } else {
        // Treat as single file path
        const fullPath = this.resolver.resolveFilePath(context.baseDir, pattern)
        const content = readFileSync(fullPath, 'utf-8')
        contents.push(content)
      }
    }

    return contents.join('\n\n')
  }
}

/**
 * Handles frontmatter extraction and parsing
 */
export class FrontmatterProcessor {
  extract(content: string): { frontmatter: any; body: string; errors: ValidationError[] } {
    const errors: ValidationError[] = []

    try {
      // Use gray-matter to extract frontmatter
      const parsed = matter(content)

      if (!parsed.data || Object.keys(parsed.data).length === 0) {
        return {
          frontmatter: {},
          body: content,
          errors: [
            {
              type: 'frontmatter',
              message: 'No frontmatter found in document',
              location: 'document start',
            },
          ],
        }
      }

      return {
        frontmatter: parsed.data,
        body: parsed.content,
        errors,
      }
    } catch (error) {
      errors.push({
        type: 'frontmatter',
        message: `Failed to extract frontmatter: ${error}`,
        location: 'document start',
      })
      return { frontmatter: {}, body: content, errors }
    }
  }
}
