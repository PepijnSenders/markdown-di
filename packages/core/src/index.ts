import matter from 'gray-matter';
import type { ProcessOptions, ProcessResult, ProcessingContext, ValidationError } from './types';
import { PartialValidator } from './validator';
import { DependencyResolver, CircularDependencyDetector } from './resolver';
import { ContentProcessor, FrontmatterProcessor } from './processor';
import { SchemaValidator } from './schema';
import { z } from 'zod';

/**
 * Main class for markdown dependency injection
 */
export class MarkdownDI {
  private partialValidator = new PartialValidator();
  private frontmatterProcessor = new FrontmatterProcessor();
  private schemaRegistry = new Map<string, z.ZodSchema<any>>();

  /**
   * Register a schema by name
   */
  registerSchema(name: string, schema: z.ZodSchema<any>): void {
    this.schemaRegistry.set(name, schema);
  }

  /**
   * Register schemas from configuration object
   */
  registerSchemas(schemas: Record<string, z.ZodSchema<any>>): void {
    Object.entries(schemas).forEach(([name, schema]) => {
      this.registerSchema(name, schema);
    });
  }

  /**
   * Get a registered schema by name
   */
  getSchema(name: string): z.ZodSchema<any> | undefined {
    return this.schemaRegistry.get(name);
  }

  /**
   * Process markdown content with dependency injection
   */
  async process(options: ProcessOptions): Promise<ProcessResult> {
    const context: ProcessingContext = {
      baseDir: options.baseDir,
      mode: options.mode || 'build',
      visitedFiles: new Set(),
      currentFile: options.currentFile
    };

    // Extract and validate frontmatter
    let { frontmatter, body, errors: frontmatterErrors } = this.frontmatterProcessor.extract(options.content);

    // Combine all errors
    const allErrors: ProcessResult['errors'] = [...frontmatterErrors];

    // Validate against schema if specified in frontmatter
    let validationSchema: z.ZodSchema<any> | null = null;

    if (frontmatter.schema && typeof frontmatter.schema === 'string') {
      const registeredSchema = this.getSchema(frontmatter.schema);

      if (!registeredSchema) {
        allErrors.push({
          type: 'schema',
          message: `Schema "${frontmatter.schema}" not found in registry. Register it with registerSchema() first.`,
          location: 'frontmatter.schema'
        });
      } else {
        const schemaValidator = new SchemaValidator({
          schema: registeredSchema,
          extend: true
        });
        const schemaResult = schemaValidator.validate(frontmatter);

        if (!schemaResult.valid) {
          allErrors.push(...schemaResult.errors);
        } else {
          // Update frontmatter with validated/transformed data
          frontmatter = schemaResult.data;
          validationSchema = registeredSchema;
        }
      }
    }

    // Validate partial syntax in content
    const syntaxValidationErrors = this.partialValidator.validate(body);
    allErrors.push(...syntaxValidationErrors);

    // Validate that injected variables are defined in frontmatter and schema
    const injectionValidationErrors = this.validateInjectionVariables(body, frontmatter, validationSchema || undefined);
    allErrors.push(...injectionValidationErrors);

    // Initialize resolver and detector
    const resolver = new DependencyResolver(context);
    const circularDetector = new CircularDependencyDetector();

    // Process content
    const processor = new ContentProcessor(resolver, circularDetector);
    const { processedContent, errors: processingErrors, dependencies } = await processor.process(
      body,
      frontmatter,
      context
    );
    allErrors.push(...processingErrors);

    // Reassemble the document with processed frontmatter
    let finalContent;
    try {
      finalContent = this.reassembleDocument(frontmatter, processedContent);
    } catch (error) {
      allErrors.push({
        type: 'frontmatter',
        message: `Failed to reassemble document: ${error}`,
        location: 'document'
      });
      finalContent = options.content; // Return original content if reassembly fails
    }

    return {
      content: finalContent,
      frontmatter,
      errors: allErrors,
      dependencies
    };
  }

  /**
   * Validate that partials exist in frontmatter
   * Mustache will handle variable validation naturally during rendering
   */
  private validateInjectionVariables(body: string, frontmatter: any, validationSchema?: z.ZodSchema<any>): ValidationError[] {
    const errors: ValidationError[] = [];

    // Extract all {{partials.xxx}} references and validate they exist
    const referenceRegex = /\{\{partials\.([^}]+)\}\}/g;
    const matches: string[] = [];
    let match;

    while ((match = referenceRegex.exec(body)) !== null) {
      matches.push(match[1].trim());
    }

    // Check each partial reference
    matches.forEach(partialKey => {
      if (!frontmatter.partials || !frontmatter.partials.hasOwnProperty(partialKey)) {
        errors.push({
          type: 'injection',
          message: `Partial '{{partials.${partialKey}}}' is not defined in frontmatter`,
          location: 'content'
        });
      }
    });

    // If there's a validation schema, ensure all required fields are present
    if (validationSchema) {
      try {
        const result = validationSchema.safeParse(frontmatter);
        if (!result.success) {
          result.error.errors.forEach((issue: any) => {
            errors.push({
              type: 'schema',
              message: `Schema validation failed: ${issue.message}`,
              location: issue.path.join('.')
            });
          });
        }
      } catch (error) {
        errors.push({
          type: 'schema',
          message: `Schema validation error: ${error instanceof Error ? error.message : String(error)}`,
          location: 'schema'
        });
      }
    }

    return errors;
  }

  /**
   * Validate markdown content without processing
   */
  async validate(options: Omit<ProcessOptions, 'mode'>): Promise<ProcessResult> {
    return this.process({ ...options, mode: 'validate' });
  }

  /**
   * Reassemble document with frontmatter and processed body
   */
  private reassembleDocument(frontmatter: any, body: string): string {
    // Filter frontmatter based on output-frontmatter field if present
    let outputFrontmatter = frontmatter;

    if (frontmatter['output-frontmatter'] && Array.isArray(frontmatter['output-frontmatter'])) {
      const allowedFields = frontmatter['output-frontmatter'];
      outputFrontmatter = {};

      // Only include fields that are in the output-frontmatter list
      for (const field of allowedFields) {
        // Don't include output-frontmatter itself
        if (field !== 'output-frontmatter' && frontmatter.hasOwnProperty(field)) {
          outputFrontmatter[field] = frontmatter[field];
        }
      }
    }

    return matter.stringify(body, outputFrontmatter);
  }
}

// Export types and classes
export * from './types';
export * from './validator';
export * from './resolver';
export * from './processor';
