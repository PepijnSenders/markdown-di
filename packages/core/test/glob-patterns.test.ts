import { describe, test, expect, beforeAll, afterAll } from 'bun:test';
import { MarkdownDI } from '../src/index';
import { mkdirSync, writeFileSync, rmSync } from 'fs';
import { join } from 'path';

const TEST_DIR = './test-glob-fixtures';

describe('Glob Pattern Support', () => {
  beforeAll(() => {
    // Create test directory structure
    mkdirSync(TEST_DIR, { recursive: true });
    mkdirSync(join(TEST_DIR, 'guides'), { recursive: true });
    mkdirSync(join(TEST_DIR, 'sections'), { recursive: true });

    // Create guide files
    writeFileSync(
      join(TEST_DIR, 'guides', 'getting-started.md'),
      '## Getting Started\n\nThis is the getting started guide.'
    );
    writeFileSync(
      join(TEST_DIR, 'guides', 'advanced.md'),
      '## Advanced Topics\n\nThis covers advanced usage.'
    );
    writeFileSync(
      join(TEST_DIR, 'guides', 'faq.md'),
      '## FAQ\n\nFrequently asked questions.'
    );

    // Create section files
    writeFileSync(
      join(TEST_DIR, 'sections', 'intro.md'),
      '# Introduction\n\nWelcome to the docs.'
    );
  });

  afterAll(() => {
    rmSync(TEST_DIR, { recursive: true, force: true });
  });

  test('should support single glob pattern (no array)', async () => {
    const content = `---
title: All Guides
partials:
  guides: guides/*.md
---

# {{title}}

{{partials.guides}}
`;

    const mdi = new MarkdownDI();
    const result = await mdi.process({
      content,
      baseDir: TEST_DIR,
    });

    expect(result.errors).toMatchSnapshot();
    expect(result.content).toMatchSnapshot();
  });

  test('should support array of glob patterns', async () => {
    const content = `---
title: Documentation
partials:
  allDocs:
    - sections/*.md
    - guides/*.md
---

# {{title}}

{{partials.allDocs}}
`;

    const mdi = new MarkdownDI();
    const result = await mdi.process({
      content,
      baseDir: TEST_DIR,
    });

    expect(result.errors).toMatchSnapshot();
    expect(result.content).toMatchSnapshot();
  });

  test('should support single file path (no glob)', async () => {
    const content = `---
title: Intro Only
partials:
  intro: sections/intro.md
---

# {{title}}

{{partials.intro}}
`;

    const mdi = new MarkdownDI();
    const result = await mdi.process({
      content,
      baseDir: TEST_DIR,
    });

    expect(result.errors).toMatchSnapshot();
    expect(result.content).toMatchSnapshot();
  });

  test('should support mixed single files and globs in array', async () => {
    const content = `---
title: Mixed Content
partials:
  content:
    - sections/intro.md
    - guides/getting-started.md
    - guides/a*.md
---

# {{title}}

{{partials.content}}
`;

    const mdi = new MarkdownDI();
    const result = await mdi.process({
      content,
      baseDir: TEST_DIR,
    });

    expect(result.errors).toMatchSnapshot();
    expect(result.content).toMatchSnapshot();
  });

  test('should handle multiple partials with different patterns', async () => {
    const content = `---
title: Multiple Partials
partials:
  intro: sections/intro.md
  guides: guides/*.md
---

# {{title}}

{{partials.intro}}

---

{{partials.guides}}
`;

    const mdi = new MarkdownDI();
    const result = await mdi.process({
      content,
      baseDir: TEST_DIR,
    });

    expect(result.errors).toMatchSnapshot();
    expect(result.content).toMatchSnapshot();
  });
});
