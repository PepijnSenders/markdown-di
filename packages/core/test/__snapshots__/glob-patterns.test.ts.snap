// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Glob Pattern Support should support single glob pattern (no array) 1`] = `[]`;

exports[`Glob Pattern Support should support single glob pattern (no array) 2`] = `
"---
title: All Guides
partials:
  guides: guides/*.md
---

# All Guides

## Advanced Topics

This covers advanced usage.

## FAQ

Frequently asked questions.

## Getting Started

This is the getting started guide.
"
`;

exports[`Glob Pattern Support should support array of glob patterns 1`] = `[]`;

exports[`Glob Pattern Support should support array of glob patterns 2`] = `
"---
title: Documentation
partials:
  allDocs:
    - sections/*.md
    - guides/*.md
---

# Documentation

# Introduction

Welcome to the docs.

## Advanced Topics

This covers advanced usage.

## FAQ

Frequently asked questions.

## Getting Started

This is the getting started guide.
"
`;

exports[`Glob Pattern Support should support single file path (no glob) 1`] = `[]`;

exports[`Glob Pattern Support should support single file path (no glob) 2`] = `
"---
title: Intro Only
partials:
  intro: sections/intro.md
---

# Intro Only

# Introduction

Welcome to the docs.
"
`;

exports[`Glob Pattern Support should support mixed single files and globs in array 1`] = `[]`;

exports[`Glob Pattern Support should support mixed single files and globs in array 2`] = `
"---
title: Mixed Content
partials:
  content:
    - sections/intro.md
    - guides/getting-started.md
    - guides/a*.md
---

# Mixed Content

# Introduction

Welcome to the docs.

## Getting Started

This is the getting started guide.

## Advanced Topics

This covers advanced usage.
"
`;

exports[`Glob Pattern Support should handle multiple partials with different patterns 1`] = `[]`;

exports[`Glob Pattern Support should handle multiple partials with different patterns 2`] = `
"---
title: Multiple Partials
partials:
  intro: sections/intro.md
  guides: guides/*.md
---

# Multiple Partials

# Introduction

Welcome to the docs.

---

## Advanced Topics

This covers advanced usage.

## FAQ

Frequently asked questions.

## Getting Started

This is the getting started guide.
"
`;
